### PR TITLE
Edit rustc_middle::dep_graph module documentation

### PR DESCRIPTION
**Pull request by pierwill**
The first commit here moves the content to `mod.rs`, the second contains edits.

The section I've called "DepNode creation and 'inference'" appears to have some outdated material. ([`DepConstructor`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/dep_graph/struct.DepConstructor.html) is a struct, not an enum.)

### CODE REVIEW

**Comment by cjgillot** [comment]
You may also want to edit the docs for `rustc_query_system::dep_graph::dep_node`, which are a copy-paste of those for `rustc_middle::dep_graph::dep_node`.

File: compiler/rustc_middle/src/dep_graph/mod.rs [hunk 1]
```diff
@@ -28,8 +30,12 @@
 //!   could not be instantiated because the current compilation session
 //!   contained no `DefId` for thing that had been removed.
 //!
+//! ## `DepNode` creation and "inference"
+//!
+// Where is the `define_dep_nodes!()` macro?
```
Comment (cjgillot) [inline compiler/rustc_middle/src/dep_graph/mod.rs:35]: The define_dep_nodes macro is defined in dep_node.rs.
It defines the `DepKind` enum, and `DepConstructor`.
`DepNode` is defined in rustc_query_system::dep_graph::dep_node.

File: compiler/rustc_middle/src/dep_graph/mod.rs [hunk 2]
```diff
@@ -28,8 +30,12 @@
...
+// How much of this section is outdated?
 //! `DepNode` definition happens in the `define_dep_nodes!()` macro. This macro
-//! defines the `DepKind` enum and a corresponding `DepConstructor` enum. The
+//! defines the `DepKind` enum and a corresponding [`DepConstructor`] enum. The
```
Comment (cjgillot) [inline compiler/rustc_middle/src/dep_graph/mod.rs:38]: Drive-by: DepConstructor should really just be a `mod`, since all it does is provide a namespace for functions constructing `DepNode`s.

File: compiler/rustc_middle/src/dep_graph/mod.rs [hunk 3]
```diff
@@ -50,6 +56,12 @@
 //! than a zeroed out fingerprint. More generally speaking, it relieves the
```
Comment (cjgillot) [inline compiler/rustc_middle/src/dep_graph/mod.rs:56]: I am not sure the API is that restricted any more. There is an escape hatch `DepNode::new_no_params` which is used in non-incremental runs.

**Comment by jyn514** [comment]
> You may also want to edit the docs for rustc_query_system::dep_graph::dep_node, which are a copy-paste of those for rustc_middle::dep_graph::dep_node.

Maybe that module should link to `rustc_middle` instead? Or use `doc = include_str!` to make sure they're kept in sync?

**Comment by pierwill** [comment]
Not at all sure I did this rebase correctly... :thinking:
